pkg/controller/imagelistpulljob/imagelistpulljob_controller.go

```diff
@@ -0,0 +1,422 @@
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// 6. Update status
+	if !util.IsJSONObjectEqual(&job.Status, newStatus) {
+		if r.updateStatus(job, newStatus); err != nil {
+			return reconcile.Result{}, fmt.Errorf("update ImageListPullJob status error: %v", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileImageListPullJob) updateStatus(job *appsv1alpha1.ImageListPullJob, newStatus *appsv1alpha1.ImageListPullJobStatus) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		imageListPullJob := &appsv1alpha1.ImageListPullJob{}
+		if err := r.Get(context.TODO(), types.NamespacedName{Namespace: job.Namespace, Name: job.Name}, imageListPullJob); err != nil {
+			return err
+		}
+		imageListPullJob.Status = *newStatus
+		return r.Status().Update(context.TODO(), imageListPullJob)
+	})
+}
+
+func (r *ReconcileImageListPullJob) computeImagePullJobActions(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) ([]*appsv1alpha1.ImagePullJob, []*appsv1alpha1.ImagePullJob) {
+	var needToDelete, needToCreate []*appsv1alpha1.ImagePullJob
+	//1. need to create
+	images, needToDelete := r.filterImagesAndImagePullJobs(job, imagePullJobs)
+	needToCreate = r.newImagePullJobs(job, images)
+	// some images delete from ImageListPullJob.Spec.Images
+	for image, imagePullJob := range imagePullJobs {
+		if !slice.ContainsString(job.Spec.Images, image, nil) {
+			needToDelete = append(needToDelete, imagePullJob)
+		}
+	}
+	return needToCreate, needToDelete
+}
+
+func (r *ReconcileImageListPullJob) calculateStatus(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) *appsv1alpha1.ImageListPullJobStatus {
+	var active, completed, succeeded int32
+	// record the failed image status
+	var failedImageStatuses []*appsv1alpha1.FailedImageStatus
+
+	for _, imagePullJob := range imagePullJobs {
+		if imagePullJob.Status.StartTime == nil {
+			continue
+		}
+
+		if imagePullJob.Status.Active > 0 {
+			active = active + 1
+		}
+
+		if imagePullJob.Status.Desired == (imagePullJob.Status.Failed + imagePullJob.Status.Succeeded) {
+			completed = completed + 1
+			//
+			if imagePullJob.Status.Failed > 0 {
+				failedImagePullJobStatus := &appsv1alpha1.FailedImageStatus{
+					ImagePullJob: imagePullJob.Name,
+					Name:         imagePullJob.Spec.Image,
+					Message:      fmt.Sprintf("Please check for details which nodes failed by using %s.", imagePullJob.Name),
+				}
+				failedImageStatuses = append(failedImageStatuses, failedImagePullJobStatus)
+			}
+
+		}
+
+		if imagePullJob.Status.Desired == imagePullJob.Status.Succeeded {
+			succeeded = succeeded + 1
+		}
+	}
+
+	// 4. status
+	newStatus := &appsv1alpha1.ImageListPullJobStatus{
+		Desired:             int32(len(job.Spec.Images)),
+		Active:              active,
+		Completed:           completed,
+		Succeeded:           succeeded,
+		Phase:               fmt.Sprintf("%d/%d", succeeded, completed),
```

Comment:
phase field seems unnecessary (completed and succeeded can be found in other fields ), i suggest we delete the field in the status and add them back later if more meaningful data is required

Reply:
It makes sense ,  and i add `Completed` field in the print column.